suppressible-error-prone/src/main/java/com/palantir/suppressibleerrorprone/SuppressingFix.java

```diff
@@ -17,120 +17,40 @@
 
 import com.google.common.collect.ImmutableSet;
-import com.google.common.collect.Iterables;
-import com.google.common.collect.Range;
 import com.google.errorprone.fixes.Fix;
 import com.google.errorprone.fixes.Replacement;
 import com.google.errorprone.fixes.Replacements.CoalescePolicy;
-import com.google.errorprone.fixes.SuggestedFix;
 import com.sun.source.tree.AnnotationTree;
-import com.sun.source.tree.AssignmentTree;
-import com.sun.source.tree.ExpressionTree;
-import com.sun.source.tree.LiteralTree;
-import com.sun.source.tree.NewArrayTree;
 import com.sun.source.tree.Tree;
 import com.sun.tools.javac.tree.EndPosTable;
 import com.sun.tools.javac.tree.JCTree.JCCompilationUnit;
-import com.sun.tools.javac.util.JCDiagnostic.DiagnosticPosition;
 import java.util.HashSet;
-import java.util.List;
 import java.util.Optional;
 import java.util.Set;
-import java.util.function.Supplier;
-import java.util.stream.Collectors;
-import java.util.stream.Stream;
+import java.util.function.Function;
 
 final class SuppressingFix implements Fix {
-    private final Optional<CharSequence> sourceCode;
-    private final Optional<? extends AnnotationTree> suppressWarnings;
-    private final Tree tree;
-    private final Set<String> errors = new HashSet<>();
+    private final Set<String> newSuppressions = new HashSet<>();
+    private final Function<EndPosTable, ImmutableSet<Replacement>> replacement;
 
     SuppressingFix(Optional<CharSequence> sourceCode, Optional<? extends AnnotationTree> suppressWarnings, Tree tree) {
-        this.sourceCode = sourceCode;
-        this.suppressWarnings = suppressWarnings;
-        this.tree = tree;
+        // See note in SuppressingReplacement about when we have to calculate stuff
+        // In order for SuppressingReplacement to calculate source code positions elements when it's constructed, it
+        // needs an EndPosTable. However, we don't get the EndPosTable until getReplacements is called. So we have
+        // to use this FirstTimeMemoizingFunction thing, that will allow use to defer creating the Replacement until
+        // we have access to the EndPosTable, then keep hold of the created SuppressingReplacement. We only need a
+        // single instance of EndPosTable to evaluate the source positions exactly once, so this works out.
+        this.replacement = new FirstTimeMemoizingFunction<>((EndPosTable endPositions) -> ImmutableSet.of(
+                new SuppressingReplacement(endPositions, newSuppressions, sourceCode, suppressWarnings, tree)));
     }
 
-    public void suppressError(String error) {
-        errors.add(error);
+    public void addSuppression(String suppression) {
+        newSuppressions.add(suppression);
     }
 
     @Override
     public ImmutableSet<Replacement> getReplacements(EndPosTable endPositions) {
-        return ImmutableSet.of(
-                new SuppressingReplacement(() -> Iterables.getOnlyElement(fix().getReplacements(endPositions))));
```

Comment:
so if I've understood correctly, `getReplacements` can be called multiple times and in effect you'll keep making thunks?

I suppose why not just use a `ConcurrentMap`? or just `Map`? with a `computeIfAbsent` method?

Reply:
If it's not the above, I think I'm not fully understanding the subtlety. From what you described in person, I'm not sure how this particular chunk in this diff makes it work 😬

Reply:
I thought about it, but then I have to save all the `sourceCode`, `tree`, `suppressWarnings` as fields rather than just consuming them in the constructor, which I think adds a bit more clutter (could hide it behind a Function - but that also feels a bit much and I'd end up wrapping it up in a type - which is basically `FirstTimeMemoizingFunction`). Once you've made the type feels a bit wasteful to make a whole map when there's just one value.

Reply:
> `getReplacements` can be called multiple times and in effect you'll keep making thunks?

`getReplacements` gets called multiple times, but only ever have one thunk/`SuppressingReplacement` thanks to the `FirstTimeMemoizingFunction`